File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode2257CountUnguardedCellsInTheGrid.cpp
#include "LeetCode2257CountUnguardedCellsInTheGrid.h"

#include <iostream>
#include <queue>
using namespace std;

int LeetCode2257CountUnguardedCellsInTheGrid::countUnguarded(int m, int n, vector<vector<int>>& guards, vector<vector<int>>& walls)
{
    //gaurd 1
    //wall 2
    //visited 3
    vector<vector<int>> grid(m, vector<int>(n, 0));
    for (auto& wall : walls)
    {
        grid[wall[0]][wall[1]] = 2;
    }
    vector<pair<int, int>> directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    int newRow{};
    int newCol{};
    for (auto& guard : guards)
    {
        grid[guard[0]][guard[1]] = 1;
    }
    for (auto& guard : guards)
    {
        for (auto& direction : directions)
        {
            newRow = guard[0];
            newCol = guard[1];
            while (true)
            {
                newRow += direction.first;
                if (newRow<0 || newRow>=m)
                    break;
                newCol += direction.second;
                if (newCol<0 || newCol>=n)
                    break;
                if (grid[newRow][newCol] == 2 || grid[newRow][newCol] == 1)
                    break;
                grid[newRow][newCol] = 3;
            }
        }
    }
    int notVisitedCount{};
    for (int i=0;i<m;i++)
    {
        for (int j=0;j<n;j++)
        {
            if (grid[i][j] == 0)
                notVisitedCount++;
        }
    }
    return notVisitedCount;
}

void LeetCode2257CountUnguardedCellsInTheGrid::Run()
{
    int m=4;
    int n=6;
    vector<vector<int>> guards={{0,0},{1,1},{2,3}};
    vector<vector<int>> walls={{0,1},{2,2},{1,4}};
    cout<<countUnguarded(m,n,guards,walls)<<'\n';
}

File: LeetCodeProblemsGo/LeetCode2257CountUnguardedCellsInTheGrid/LeetCode2257CountUnguardedCellsInTheGrid.go
package LeetCode2257CountUnguardedCellsInTheGrid

func countUnguarded(m int, n int, guards [][]int, walls [][]int) int {

	// 1 wall
	//2 guard
	// 3 supervised
	grid := make([][]int, m)
	for i := 0; i < m; i++ {
		grid[i] = make([]int, n)
	}
	for _, wall := range walls {
		grid[wall[0]][wall[1]] = 1
	}
	for _, guard := range guards {
		grid[guard[0]][guard[1]] = 2
	}
	directions := [][]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
	newRow := 0
	newCol := 0
	for _, guard := range guards {
		for _, direction := range directions {
			newRow = guard[0]
			newCol = guard[1]
			for true {
				newRow += direction[0]
				if newRow < 0 || newRow >= m {
					break
				}
				newCol += direction[1]
				if newCol < 0 || newCol >= n {
					break
				}
				if grid[newRow][newCol] == 1 || grid[newRow][newCol] == 2 {
					break
				}
				grid[newRow][newCol] = 3
			}
		}
	}
	result := 0
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] == 0 {
				result++
			}
		}
	}
	return result
}
